lb_content_resolver/metadata_lookup.py

```diff
@@ -0,0 +1,123 @@
+                    return
+                args = []
+                mbid_to_id_index = {}
+
+        if len(args) > 0:
+            self.lookup_chunk(args, mbid_to_id_index)
+
+
+    def lookup_chunk(self, args, mbid_to_id_index):
+        """
+            This function carries out the actual lookup of the metadata and inserting the
+            popularity and tags into the DB for the given chunk of recordings.
+        """
+
+        r = requests.post("https://labs.api.listenbrainz.org/bulk-tag-lookup/json", json=args)
+        if r.status_code != 200:
+            print("Fail: %d %s" % (r.status_code, r.text))
+            return False
+
+        recording_pop = {}
+        recording_tags = {}
+        tags = set()
+        for row in r.json():
+            mbid = str(row["recording_mbid"])
+            recording_pop[mbid] = row["percent"]
+            if mbid not in recording_tags:
+                recording_tags[mbid] = { "artist": [], "release-group": [], "recording": [] }
+
+            recording_tags[mbid][row["source"]].append(row["tag"])
+            tags.add(row["tag"])
+
+        tags = list(tags)
+        with db.atomic():
+
+            # This DB code is pretty messy -- things I take for granted with Postgres are not
+            # available in SQLite or the PeeWee ORM. But, this might be ok, since we're not 
+            # updating millions of rows constantly.
+
+            # First update recording_metadata table
+            mbids = recording_pop.keys()
+            for mbid in list(set(mbids)):
+                mbid = str(mbid)
+                row = mbid_to_id_index[mbid]
+                if row[3] is None:
+                    recording_metadata = RecordingMetadata.create(recording=row[0],
+                                                                  popularity=recording_pop[mbid],
+                                                                  last_updated=datetime.datetime.now())
+                    recording_metadata.save()
+                else:
+                    recording_metadata = RecordingMetadata.replace(id=row[2],
+                                                                  recording=row[0],
+                                                                  popularity=recording_pop[mbid],
+                                                                  last_updated=datetime.datetime.now())
+
+                    recording_metadata.execute()
+
+            # Next delete recording_tags
+            mbids = recording_tags.keys()
+            for mbid in mbids:
+                db.execute_sql("""DELETE FROM recording_tag WHERE recording_id in (
+                                       SELECT id FROM recording WHERE recording_mbid = ?
+                                  )""", (mbid,))
+
+            # Finally, insert new recording tags
+
+            # Sqlite doesn't seem to support RETURNING with OR IGNORE, so we have to do this in two steps.
```

Comment:
This is the same behavior as PostgreSQL, ON CONFLICT DO NOTHING doesn't return existing rows. You can work around it by using ON CONFLICT DO UPDATE, even though the update won't change anything in the table but because an UPDATE was performed on the data in the disk, those rows will be returned. Sqlite seems to support that too: https://www.sqlite.org/draft/lang_UPSERT.html

Reply:
Thanks for the tip!